uc-crux-llvm/src/UCCrux/LLVM/Run/Check.hs

```diff
@@ -0,0 +1,406 @@
+              })
+  where
+    overrides ::
+      IsSymInterface sym =>
+      HasLLVMAnn sym =>
+      IO [ ( Sim.SymCreateOverrideFn sym arch
+           , SomeCheckedCalls m sym arch
+           )
+         ]
+    overrides =
+      for
+        (Map.toList contracts)
+        (\(func, Some (TypedConstraints tcs types)) ->
+           do CFGWithTypes ovCfg argFTys _retTy _varArgs <-
+                pure (findFun modCtx (FuncDefnSymbol func))
+              let argCTys = Crucible.cfgArgTypes ovCfg
+              ovFunCtx <-
+                case makeFunctionContext modCtx func argFTys argCTys of
+                  Left err ->
+                    panic
+                      "checkInferredContracts"
+                      [Text.unpack (ppFunctionContextError err)]
+                  Right funCtx' -> return funCtx'
+              let ?memOpts = CruxLLVM.memOpts llOpts
+              case testEquality argFTys types of
+                Nothing -> panic "checkInferredContracts" []
+                Just Refl ->
+                  do ref <- IORef.newIORef []
+                     return
+                       ( Sim.SymCreateOverrideFn $
+                           \_sym ->
+                             return $
+                               Check.createCheckOverride
+                                 appCtx
+                                 modCtx
+                                 ref
+                                 types
+                                 tcs
+                                 ovCfg
+                                 (FuncDefnSymbol func)
+                       , SomeCheckedCalls
+                           (\f -> f func ovFunCtx =<< IORef.readIORef ref)
+                       )
+        )
+
+-- | Postcondition: The keys of the returned 'Map' are exactly the
+-- 'EntryPoints'.
+checkInferredContractsFromEntryPoints ::
+  forall m arch msgs.
+  Crux.Logs msgs =>
+  Crux.SupportsCruxLogMessage msgs =>
+  ArchOk arch =>
+  AppContext ->
+  ModuleContext m arch ->
+  HandleAllocator ->
+  CruxOptions ->
+  LLVMOptions ->
+  -- | Where to begin symbolic execution
+  EntryPoints m ->
+  -- | Inferred function contracts
+  Map (DefnSymbol m) (Some (TypedConstraints m)) ->
+  IO (Map (DefnSymbol m) (SomeCheckResult m arch))
+checkInferredContractsFromEntryPoints appCtx modCtx halloc cruxOpts llOpts entries contracts =
+  fmap Map.fromList $
+    for (getEntryPoints entries) $
+      \entry ->
+        do CFGWithTypes cfg argFTys _retTy _varArgs <-
+             pure (findFun modCtx (FuncDefnSymbol entry))
+
+           funCtx <-
+             case makeFunctionContext modCtx entry argFTys (Crucible.cfgArgTypes cfg) of
+               Left err ->
+                 panic
+                   "checkInferredContracts"
+                   [Text.unpack (ppFunctionContextError err)]
+               Right funCtxF -> return funCtxF
+           result <-
+            checkInferredContracts
```

Comment:
Does it matter that there are multiple `IORef`s tracking checked functions under here? Is it necessary that there are multiple?

Reply:
When you say "there are multiple `IORef`s", do you mean the `IORef`s that are passed to `createCheckOverride`?

If so, then I think it's fine in the sense that `checkInferredContracts` returns a way to read each of them (obliquely, through a `SomeCheckedCalls`).

> Is it necessary that there are multiple?

I think it makes the interface to `createCheckOverride` nicer if the results can depend on the (type-level parameters that indicate) types of the arguments to the function. Since these functions might all have different, non-statically-known types, we have to then have one `IORef` per check override.

Reply:
Okay, that makes sense to me. Yes - I was thinking of that set of IORefs